edx/edx-arch-experiments
# feat!: Detect anomalous Datadog traces; upgrade to Python 3.11

- Detect and log anomalous traces in DatadogDiagnosticMiddleware.
- Drop Python 3.8 support and change to Python 3.11 minimum due to testing needs.
- Add tests for existing middleware code

Manual testing:

1. Install in devstack, add middleware to lms/envs/devstack.py, and
   enable Waffle flag.
2. Make some requests and observe lack of new logging.
3. Add `self.dd_tracer.current_root_span().finish()` to start of
   `process_view` (faking an anomalous trace) and reinstall in devstack
4. Make some more requests and observe log output like the following:
   `Anomalous Datadog local root span (duration already set): id = 66abf66b000000001168265a391d04a3; duration = 0.108 sec; worker age = 1.560 sec`

Unit test improvements:

- Install ddtrace at the test.in level
- Add tests for new and existing middleware functionality
- Test with Python 3.11, not 3.8. This isn't just because we're upgrading
  to 3.11 across our repos, but also because I ran into a really weird bug
  with ddtrace and/or jsonschema that only occurs with Python 3.8:
  https://github.com/DataDog/dd-trace-py/issues/9950

----

**Merge checklist:**
Check off if complete *or* not applicable:
